src/hooks/cgroups/pbs_cgroups.PY

```diff
@@ -4451,7 +4501,40 @@ class CgroupUtils(object):
         pbs.logmsg(pbs.EVENT_DEBUG4, '%s: Setting cgroup limits for: %s' %
                    (caller_name(), hostresc))
         # The vmem limit must be set after the mem limit, so sort the keys
+        # also ensures we get to cpuset.cpus before cpuset.mem
+        # important for zero cpu jobs migrated to root cpuset
         for resc in sorted(hostresc):
+            if resc == 'ncpus':
+                # In case of HT, may need to multiply the ncpus;
+                # done here since "node" is not passed down to set_limit
+                # This only sets the cpu controller limits/shares,
+                # since the cpuset stuff is done through
+                # cpuset.cpus and cpuset.mems
+                htpc = 1
+                if (self.cfg['cgroup']['cpu']['enabled']
+                        and self.cfg['use_hyperthreads']
+                        and self.cfg['ncpus_are_cores']):
+                    if 'hyperthreads_per_core' in node.cpuinfo:
+                        htpc = node.cpuinfo['hyperthreads_per_core']
+                self.set_limit(resc, hostresc[resc] * htpc, jobid)
+            elif ((resc == "cpuset.cpus")
+                  and not hostresc['cpuset.cpus']
+                  and self.cfg['cgroup']['cpuset']['allow_zero_cpus']):
+                # zero cpus jobs -- delete the cpuset made for the job,
+                # allowing the job processes to be in the root cpuset
+                path = self._cgroup_path('cpuset', '', jobid)
+                success = self._remove_cgroup(path)
+                if not success:
+                    pbs.logmsg(pbs.EVENT_DEBUG,
+                               '%s: Failed to delete cpuset %s'
+                               ' for zero cpus job %s'
+                               % (caller_name(), path, jobid))
+                    pbs.event().reject('Failed to delete cpuset'
+                                       ' for zero cpus job')
+                    # Hopefully the orphan cleaner will eventually
+                    # get rid of it
+            else:
+                # For everything else just pass it down
```

Comment:
Either remove the else or add a call to pass here.

Reply:
In my original copy the self.set_limit behind this was indented under the else...let me check the logic.

Reply:
Yup, somehow an indentation was lost in translation when some things were merged. Go figure.

The logic is that seld.setlimit needs _NOT_ to be done for the two special cases. For "ncpus" there is a set_limit that takes care of multiplying what we need to set by the number of threads if necessarey; for cpuset.cpus we need not call set_limit for ncpus==0 (in that case hostresc['cpuset.cpus'] is empty, and we test for that), but delete the cpuset (after which the processes will be added to the main pbspro.slice cpuset).

In all *other* cases, _then_ we "just" have to call setlimit.